lightning/src/ln/channelmanager.rs

```diff
@@ -2293,10 +2316,37 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 								let mut payment_secrets = self.pending_inbound_payments.lock().unwrap();
 								match payment_secrets.entry(payment_hash) {
 									hash_map::Entry::Vacant(_) => {
-										log_trace!(self.logger, "Failing new HTLC with payment_hash {} as we didn't have a corresponding inbound payment.", log_bytes!(payment_hash.0));
-										fail_htlc!(claimable_htlc);
+										match claimable_htlc.claim_type {
+											ClaimableType::Invoice(_) => {
+												log_trace!(self.logger, "Failing new HTLC with payment_hash {} as we didn't have a corresponding inbound payment.", log_bytes!(payment_hash.0));
+												fail_htlc!(claimable_htlc);
+											},
+											ClaimableType::Spontaneous(preimage) => {
+												let htlc_vec = channel_state.claimable_htlcs.entry(payment_hash)
+													.or_insert(Vec::new());
+												if htlc_vec.len() != 0 {
+													log_trace!(self.logger, "Failing new keysend HTLC with payment_hash {} for a duplicative payment hash", log_bytes!(payment_hash.0));
+													fail_htlc!(claimable_htlc);
+												} else {
+													htlc_vec.push(claimable_htlc);
+													new_events.push(events::Event::PaymentReceived {
+														payment_hash,
+														amt: amt_to_forward,
+														purpose: events::PaymentPurpose::SpontaneousPayment(preimage),
+													});
+												}
```

Comment:
This is a little difficult to follow. Might be easier to read if you match on `entry` and place the success and failure cases in the `Vacant` and `Occupied` cases, respectively, if possible.